pkg/astnormalization/implicit_extend_root_operation.go

```diff
@@ -0,0 +1,57 @@
+package astnormalization
+
+import (
+	"github.com/jensneuse/graphql-go-tools/pkg/ast"
+	"github.com/jensneuse/graphql-go-tools/pkg/astvisitor"
+)
+
+/*
+type Query {...}
+type Mutation {...}
+type Subscription {...}
+
+will be,
+
+extend type Query {...}
+extend type Mutation {...}
+extend type Subscription {...}
+
+this also works if root types are defined in schema{...} with other names.
+root types are left unmodified if they have no fields, directives or implements any interface.
+*/
+type implicitExtendRootOperationVisitor struct {
+	operation *ast.Document
+}
+
+func implicitExtendRootOperation(walker *astvisitor.Walker) {
+	v := &implicitExtendRootOperationVisitor{}
+	walker.RegisterEnterDocumentVisitor(v)
+	walker.RegisterEnterObjectTypeDefinitionVisitor(v)
+}
+
+func (v *implicitExtendRootOperationVisitor) EnterDocument(operation, _ *ast.Document) {
+	v.operation = operation
+}
+
+func (v *implicitExtendRootOperationVisitor) EnterObjectTypeDefinition(ref int) {
+	node := v.operation.ObjectTypeDefinitions[ref]
+	if !(node.HasDirectives || node.HasFieldDefinitions || len(node.ImplementsInterfaces.Refs) > 0) {
+		return
+	}
+	switch v.operation.Input.ByteSliceString(node.Name) {
+	case implicitQueryTypeName, implicitMutationTypeName, implicitSubscriptionTypeName,
+		v.operation.Index.QueryTypeName.String(), v.operation.Index.MutationTypeName.String(), v.operation.Index.SubscriptionTypeName.String():
+		for i := range v.operation.RootNodes {
+			if v.operation.RootNodes[i].Ref == ref && v.operation.RootNodes[i].Kind == ast.NodeKindObjectTypeDefinition {
+				// give this node a new NodeKind of ObjectTypeExtension
+				var newKind ast.ObjectTypeExtension
+				newKind.ObjectTypeDefinition = node
+				v.operation.ObjectTypeExtensions = append(v.operation.ObjectTypeExtensions, newKind)
```

Comment:
This could be extracted into a helper on a ast.Document (`pkg/ast/ast_object_type_extension.go`) to make it consistent with existing code base

```go
func (d *Document) AddObjectTypeDefinitionExtension(extension ObjectTypeExtension) (ref int) {
	d. ObjectTypeExtensions = append(d. ObjectTypeExtensions, extension)
	return len(d.ObjectTypeExtensions) - 1
}
```

and u will be able to use it like this

```go
newRef := v.operation.AddObjectTypeDefinitionExtension(ast.ObjectTypeExtension{ObjectTypeDefinition: node})
```